[py2py3] migration of DQMUpload.py

mapellidario opened the pull request:
Fixes #9926

#### Status
Ready 

#### Description

I restarted the work on this topic, driven by investigations necessary for #10651. A lot of time passed, and with time we have a better insight of what is going on here:

- [x] #9926 is caused by python-future's backport of py3's urllib to py2. This can be avoided by simply continue using urllib2 in py2. See logs at [1]
- [x] #10651 fixed a problem that arises when using py3 central services and py2 wmagent. The `url` argument in `urllib2.Request(url)` ended up being of type unicode, which mixed the types in the httplib buffer, causing a the same problem seen in #9926
- [x] I added a unittest. ~~I am making it fail because `dqm.upload()` is returning http code 500. Maybe it's just a matter of chaning the filename, but i have not looked for what it should be, yet  See logs for~~
  - ~~[py2 logs](https://cmssdt.cern.ch/dmwm-jenkins/view/All/job/DMWM-WMCore-PR-test/12064/testReport/junit/WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t/DQMUpload_t/testA_Upload/) for `WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t.testA_Upload`~~
  - ~~[py3 logs](https://cmssdt.cern.ch/dmwm-jenkins/view/All/job/DMWM-WMCore-PR-test/12064/testReport/junit/WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t/DQMUpload_t/testA_Upload_2/) for `WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t.testA_Upload`~~
  - Thanks to @amaltaro for providing the fix! :).
  - Since this unittest uploads a  144MB root file coming directly from Alan's eos to the DQM every time we run this new unittest, we decided that it is better not to run this often. However it is good to keep this around in case of bugs -> we mark this unittest as integration

The solution was to avoid using in py2 the backport of py3 urllib provided by future, which is broken. So the only "quick" solution is to select at runtime with a `try/except` clause the proper classes and functions.

An alternative would be refactoring all these machinery and start using `request` instead, but it may not be worth the effort.

#### Is it backward compatible (if not, which system it affects?)
YES

#### External dependencies / deployment changes
DQMUpload.py will continue to depend on python-future

amaltaro commented:
@mapellidario Thanks Dario. I'm going to redeploy vocms0193, apply this patch and run some tests. Once I have any results, will report them back here.

amaltaro commented:
@mapellidario as you expected, no unicode data in there:
```
2020-09-24 17:02:18,414:INFO:DQMUpload:[9926] header ('Content-length', '49159387') <type 'str'>
2020-09-24 17:02:18,415:INFO:DQMUpload:[9926] header ('Content-type', 'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------') <type 'str'>
2020-09-24 17:02:18,415:INFO:DQMUpload:[9926] header ('Accept-encoding', 'gzip') <type 'str'>
2020-09-24 17:02:18,415:INFO:DQMUpload:[9926] header ('User-agent', 'WMAgent python/2.7.13') <type 'str'>
2020-09-24 17:02:18,415:INFO:DQMUpload:[9926] Request data type: <type 'str'>
2020-09-24 17:02:18,571:ERROR:DQMUpload:HTTP upload failed! Error:
'ascii' codec can't decode byte 0xec in position 479: ordinal not in range(128)
Traceback (most recent call last):
  File "/srv/job/WMCore.zip/WMCore/WMSpec/Steps/Executors/DQMUpload.py", line 157, in httpPost
    (headers, data) = self.upload(uploadURL, args, filename)
  File "/srv/job/WMCore.zip/WMCore/WMSpec/Steps/Executors/DQMUpload.py", line 259, in upload
    result = opener.open(datareq)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/py2-future/0.18.2/lib/python2.7/site-packages/future/backports/urllib/request.py", line 500, in open
    response = self._open(req, data)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/py2-future/0.18.2/lib/python2.7/site-packages/future/backports/urllib/request.py", line 518, in _open
    '_open', req)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/py2-future/0.18.2/lib/python2.7/site-packages/future/backports/urllib/request.py", line 472, in _call_chain
    result = func(*args)
  File "/srv/job/WMCore.zip/WMCore/Services/HTTPS/HTTPSAuthHandler.py", line 58, in https_open
    return self.do_open(self.get_connection, req)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/py2-future/0.18.2/lib/python2.7/site-packages/future/backports/urllib/request.py", line 1290, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/httplib.py", line 1082, in _send_request
    self.endheaders(body)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "/cvmfs/cms.cern.ch/COMP/slc6_amd64_gcc493/external/python/2.7.13/lib/python2.7/httplib.py", line 880, in _send_output
    msg += message_body
UnicodeDecodeError: 'ascii' codec can't decode byte 0xec in position 479: ordinal not in range(128)
```

I have the suspicion that it might be coming from the marshall method. Will try to play with it tomorrow, but in case you have other suggestions, please do let me know.

vkuznet commented:
I'm not sure why you said no unicode data is here, since the error message is very clear:
```
'ascii' codec can't decode byte 0xec in position 479: ordinal not in range(128)
```
It means that there is a special character which is not in ASCII table and it should be treated (decoded/encoded) properly. There are plenty of materials on internet about it.

mapellidario commented:
I am suprised. I hoped that the non-ascii character was inserted by `marshall` and would have been in `datareq.data`, so I added `         logging.info("[9926] Request data: %s %s", datareq.data)` at line 256, but I do not see this logging in your stacktrace. Do you have any idea of why this could have been missed?

mapellidario commented:
However, what puzzles me the most is that I tried to insert some unicode in `datareq.data` in a [script here](https://gist.github.com/mapellidario/761a1a3b7dc6b70c275d2ee8aedb882b), but I have not been able to replicate this error. There is something that I am missing, and this can either be a huge mistake on my side or a subtle quirk that I missed.

vkuznet commented:
Another observation, why do you run using this architecture slc6_amd64_gcc493? All cmsweb stack is now using slc7_amd64_gcc630.

And, regarding str vs unicode, may be this will help
https://stackoverflow.com/questions/18649512/unicodedecodeerror-ascii-codec-cant-decode-byte-0xe2-in-position-13-ordinal
https://stackoverflow.com/questions/21129020/how-to-fix-unicodedecodeerror-ascii-codec-cant-decode-byte/21129492

mapellidario commented:
test this please

mapellidario commented:
Hello @amaltaro! The status of this PR is: when we try to upload a file, we do not respect a specific filename convention and we get http code 500 back. This makes me believe that urllib is douing its job and is sending the data properly, though, and makes me believe that we fixed the issue. However, I am not sure if this is a good option, so I am making the unittest fail with a `raise Exception` so that you can read the logs from jenkins.

I see two ways of proceeding here
1. We ignore the error code 500 and we get it as a success from our side, making the new unittest mean "we only care about sending the data, we do not really care about respecting all the filename conventions"
2. We fix the filename and we make sure that the data we send gets actually uploaded. This means that every time we run the unittest, we upload data to the DQM gui, maybe not something we really want though.

Do you have any suggestions?

amaltaro commented:
And I've only seen your question because I ran into issues importing urllib2 in python3, sorry.

Dario, it's not clear to me whether you are concerned only with the unit test, or with this code in "real" production mode. Nonetheless, not respecting the http request status code would be very confusing IMO.

Let me patch my test agent and get a test running while I have a look at this code.

amaltaro commented on test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py:
```diff
@@ -0,0 +1,64 @@
+        pass
+
+    def testA_Upload(self):
+        filename = "/tmp/dqm.root"
```
Ahhh, it's probably failing with the root file name because of this line, no?

mapellidario replied:
Aha! let me check it! :)

mapellidario replied:
It worked! thanks Alan! :) sometimes a pair of new eyes is worth 2h of work ahaha

mapellidario commented:
Nono, i am concerned only with always uploading the file in the unittest, i hope that real production code can upload all the files we wish :)

mapellidario commented:
@amaltaro using the name of the file that you provied instead of `dqm.root` worked indeed! i will push a new commit.

I paste below some logging of the successful request which may be useful for future-us. Notice that I am artificially making it fail just to retrieve the logs (with some debugging information) in case of a successfull request. 

### py3 log

```plaintext
#1 testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t) ... ERROR

======================================================================
ERROR: testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py", line 62, in testA_Upload
    raise Exception
Exception: 
-------------------- >> begin captured logging << --------------------
root: INFO: Steps.Executor logging started
root: INFO: Using proxy file: /home/dmwm/certs-mount/x509up_u99307
root: INFO: Using CA certificate path: None
root: INFO: HTTP POST upload arguments:
  ==> checksum: md5:776e66d494a6eeabd52b1d7fb50e9bd4
  ==> size: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: Found 142 default trusted CA certificates.
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: SSL context manager created with the following settings:
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   check_hostname : True
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   options : Options.OP_ALL|OP_NO_SSLv3|OP_NO_SSLv2|OP_CIPHER_SERVER_PREFERENCE|OP_SINGLE_DH_USE|OP_SINGLE_ECDH_USE|OP_NO_COMPRESSION
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   protocol : _SSLMethod.PROTOCOL_TLS
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_flags : VerifyFlags.VERIFY_X509_TRUSTED_FIRST
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_mode : VerifyMode.CERT_REQUIRED
root: DEBUG: encode value - <class 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode payload - <class 'bytes'>, b'md5:776e66d494a6eeabd52b1d7fb50e9bd4'
root: DEBUG: encode value - <class 'int'>, 150553620
root: DEBUG: encode payload - <class 'bytes'>, b'150553620'
root: DEBUG: encode body (without binary file) -<class 'bytes'>, b'------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="checksum"\r\n\r\nmd5:776e66d494a6eeabd52b1d7fb50e9bd4\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="size"\r\n\r\n150553620\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="file"; filename="DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"\r\nContent-Type: application/octet-stream\r\nContent-Length: 150553620\r\n'
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG: POST method to https://cmsweb-testbed.cern.ch/dqm/dev//data/put
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG:   with the following headers: {'Accept-encoding': 'gzip', 'User-agent': 'WMAgent python/3.8.2', b'Content-type': b'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------'}
root: DEBUG: headers: <class 'http.client.HTTPMessage'> Date: Fri, 13 Aug 2021 11:03:55 GMT
Server: CherryPy/5.4.0
Set-Cookie: cms-auth=a1db9d8e44eac6a056b1b6d5b0d937561bd90cef6b250f9c42967d03247917ceb85188e2ef081d14;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
Dqm-Status-Detail: Wrote 150553620 bytes
Content-Length: 376
Dqm-Status-Code: 100
Dqm-Status-Message: File saved
Content-Type: text/html;charset=utf-8
CMS-Server-Time: D=6784126 t=1628852635676700
Connection: close


root: DEBUG: data: <class 'bytes'> b'Thanks.\nThe server received your file. However this does not mean that the file was indexed successfully.\nA different server process will now check the filename to determine in which folder to archive it.\nAfter that, another process will attempt to index it. This might take some minutes.\nIn case of doubt, please check the server logs of the receive-daemon and import-daemon.'
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 1 test in 9.509s

FAILED (errors=1)
Testing complete, there are now 1 threads
```

### py2 log

```plaintext
#1 testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t) ... ERROR

======================================================================
ERROR: testA_Upload (WMCore_t.WMSpec_t.Steps_t.Executors_t.DQMUpload_t.DQMUpload_t)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py", line 62, in testA_Upload
    raise Exception
Exception: 
-------------------- >> begin captured logging << --------------------
root: INFO: Steps.Executor logging started
root: INFO: Using proxy file: /home/dmwm/certs-mount/x509up_u99307
root: INFO: Using CA certificate path: None
root: INFO: HTTP POST upload arguments:
  ==> checksum: md5:776e66d494a6eeabd52b1d7fb50e9bd4
  ==> size: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: Found 142 default trusted CA certificates.
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: SSL context manager created with the following settings:
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   check_hostname : True
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   options : 2197947391
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   protocol : 2
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_flags : 32768
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_mode : 2
root: DEBUG: encode value - <type 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode payload - <type 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode value - <type 'int'>, 150553620
root: DEBUG: encode payload - <type 'str'>, 150553620
root: DEBUG: encode body (without binary file) -<type 'str'>, ------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="checksum"

md5:776e66d494a6eeabd52b1d7fb50e9bd4
------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="size"

150553620
------------=_DQM_FILE_BOUNDARY_=-----------
Content-Disposition: form-data; name="file"; filename="DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"
Content-Type: application/octet-stream
Content-Length: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG: POST method to https://cmsweb-testbed.cern.ch/dqm/dev//data/put
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG:   with the following headers: {'Content-length': '150554231', 'Content-type': 'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------', 'Accept-encoding': 'gzip', 'User-agent': 'WMAgent python/2.7.13'}
root: DEBUG: headers: <type 'instance'> Date: Fri, 13 Aug 2021 11:03:57 GMT
Server: CherryPy/5.4.0
Set-Cookie: cms-auth=c6bbcbb76af633e7876c40bb2a02b99053c8353f76dd7bcc2c9d133f68bb4236ff48ee3796edcf3f;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
Dqm-Status-Detail: Wrote 150553620 bytes
Content-Length: 376
Dqm-Status-Code: 100
Dqm-Status-Message: File saved
Content-Type: text/html;charset=utf-8
CMS-Server-Time: D=6561068 t=1628852637711378
Connection: close

root: DEBUG: data: <type 'str'> Thanks.
The server received your file. However this does not mean that the file was indexed successfully.
A different server process will now check the filename to determine in which folder to archive it.
After that, another process will attempt to index it. This might take some minutes.
In case of doubt, please check the server logs of the receive-daemon and import-daemon.
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 1 test in 10.175s

FAILED (errors=1)
Testing complete, there are now 1 threads

```

mapellidario committed:
[py2py3] modernize DQMUpload.py

mapellidario committed:
[py2py3] modernize DQMUpload.py - test

amaltaro commented:
I am glad to hear it works now! I was hoping to have real test results, but my test workflow was still stuck in acquired...

Regarding the unit test, I think that file is way too big for many daily uploads, Dario. I would suggest to keep that unit test around - because it's going to be extremely useful if we have to test/debug things - but mark them as `@integration`, such that they do not run in jenkins.

amaltaro commented:
Jenkins is misbehaving and it's triggering a CI run after another. I'm temporarily closing this PR to stop it.

PS.: My tests jobs have not yet completed, they seem to keep running forever...

mapellidario commented:
Thanks to @amaltaro tests, we found a bug in DQMUpload.httpPost() when running in py3, caused by the following line 

https://github.com/dmwm/WMCore/blob/8b4dd2ded5f9c49d25c5672e2bb885db57687d3d/src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py#L141

the bug has been fixed in the last commit, i added a new unittest that covers that line, too.

Achtung: i removed the `@attr("integration")` flag from the two unittests only temporarily because i want to see those passing in jenkins. i will put back those flags as soon as we have a successfull report.

mapellidario commented:
As above, I made the new unittest `test_httppost` fail in order to retrieve the logs. Apart from the artificial failure, a good log from the new unittest is:

```plaintext
path lists is ['test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py', 'DQMUpload_t.test_httppost']
Out of 1 cases, we will run 1
#1 test_httppost ... ERROR

======================================================================
ERROR: test_httppost
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py", line 100, in test_httppost
    raise Exception
Exception: 
-------------------- >> begin captured logging << --------------------
root: INFO: Steps.Executor logging started
root: INFO: HTTP Upload is about to start:
 => URL: https://cmsweb-testbed.cern.ch/dqm/dev/
 => Filename: /tmp/DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root

root: INFO: Using proxy file: /home/dmwm/certs-mount/x509up_u99307
root: INFO: Using CA certificate path: None
root: INFO: HTTP POST upload arguments:
  ==> checksum: md5:776e66d494a6eeabd52b1d7fb50e9bd4
  ==> size: 150553620

WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: Found 142 default trusted CA certificates.
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO: SSL context manager created with the following settings:
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   check_hostname : True
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   options : Options.OP_ALL|OP_NO_SSLv3|OP_NO_SSLv2|OP_CIPHER_SERVER_PREFERENCE|OP_SINGLE_DH_USE|OP_SINGLE_ECDH_USE|OP_NO_COMPRESSION
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   protocol : _SSLMethod.PROTOCOL_TLS
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_flags : VerifyFlags.VERIFY_X509_TRUSTED_FIRST
WMCore.Services.HTTPS.HTTPSAuthHandler: INFO:   verify_mode : VerifyMode.CERT_REQUIRED
root: DEBUG: encode value - <class 'str'>, md5:776e66d494a6eeabd52b1d7fb50e9bd4
root: DEBUG: encode payload - <class 'bytes'>, b'md5:776e66d494a6eeabd52b1d7fb50e9bd4'
root: DEBUG: encode value - <class 'int'>, 150553620
root: DEBUG: encode payload - <class 'bytes'>, b'150553620'
root: DEBUG: encode body (without binary file) -<class 'bytes'>, b'------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="checksum"\r\n\r\nmd5:776e66d494a6eeabd52b1d7fb50e9bd4\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="size"\r\n\r\n150553620\r\n------------=_DQM_FILE_BOUNDARY_=-----------\r\nContent-Disposition: form-data; name="file"; filename="DQM_V0001_R000000001__RelValH125GGgluonfusion_13__CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1705_Validation_TEST_Alan_v67-v11__DQMIO.root"\r\nContent-Type: application/octet-stream\r\nContent-Length: 150553620\r\n'
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG: POST method to https://cmsweb-testbed.cern.ch/dqm/dev//data/put
WMCore.Services.HTTPS.HTTPSAuthHandler: DEBUG:   with the following headers: {'Accept-encoding': 'gzip', 'User-agent': 'WMAgent python/3.8.2', b'Content-type': b'multipart/form-data; boundary=----------=_DQM_FILE_BOUNDARY_=-----------'}
root: INFO: HTTP upload finished succesfully with response:
  Status code: 100
  Message: File saved
  Detail: Wrote 150553620 bytes
  Data: b'Thanks.\nThe server received your file. However this does not mean that the file was indexed successfully.\nA different server process will now check the filename to determine in which folder to archive it.\nAfter that, another process will attempt to index it. This might take some minutes.\nIn case of doubt, please check the server logs of the receive-daemon and import-daemon.'

--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
XML: /home/dmwm/wmcore_unittest/WMCore/nosetests.xml
----------------------------------------------------------------------
Ran 1 test in 8.386s

FAILED (errors=1)
Testing complete, there are now 1 threads
```

amaltaro commented:
Thanks for the prompt fix, Dario. Your last commit indeed fixed the issue and harvesting jobs are now succeeding.
Feel free to squash those commits properly, I will make a new release with these changes in by tomorrow. Thanks again!

mapellidario commented:
After jenkins reported that the two new unittests are passing both in py2 and py3, i added again the `@attr("integration")`  flag on those

amaltaro commented on test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/DQMUpload_t.py:
```diff
@@ -0,0 +1,101 @@
+    def tearDown(self):
+        pass
+
+    @attr("integration")
```
@mapellidario I think you missed the
```
from nose.plugins.attrib import attr
```

mapellidario replied:
aaaaaa you are correct, sorry! thanks for th suggestions!

amaltaro commented:
Thanks for making these changes, Dario.
This has been tested with Py3 WMAgent and it works fine. We will find out in the next replay or in the coming days whether this works for Py2 as well...
